engine/graphics/src/test/test_app_graphics.cpp

```diff
@@ -0,0 +1,497 @@
+
+        //////////// RENDER PASS ////////////
+        uint8_t sub_pass_0_color_attachments[] = { 0 };
+        uint8_t sub_pass_1_color_attachments[] = { 1 };
+        uint8_t sub_pass_1_input_attachments[] = { 0 };
+
+        dmGraphics::CreateRenderPassParams rp = {};
+        rp.m_SubPassCount = 2;
+
+        rp.m_SubPasses[0].m_ColorAttachmentIndices      = sub_pass_0_color_attachments;
+        rp.m_SubPasses[0].m_ColorAttachmentIndicesCount = DM_ARRAY_SIZE(sub_pass_0_color_attachments);
+
+        rp.m_SubPasses[1].m_ColorAttachmentIndices      = sub_pass_1_color_attachments;
+        rp.m_SubPasses[1].m_ColorAttachmentIndicesCount = DM_ARRAY_SIZE(sub_pass_1_color_attachments);
+
+        rp.m_SubPasses[1].m_InputAttachmentIndices      = sub_pass_1_input_attachments;
+        rp.m_SubPasses[1].m_InputAttachmentIndicesCount = DM_ARRAY_SIZE(sub_pass_1_input_attachments);
+
+        rp.m_DependencyCount = 3;
+        rp.m_Dependencies[0].m_Src = dmGraphics::SUBPASS_EXTERNAL;
+        rp.m_Dependencies[0].m_Dst = 0;
+
+        rp.m_Dependencies[1].m_Src = 0;
+        rp.m_Dependencies[1].m_Dst = 1;
+
+        rp.m_Dependencies[2].m_Src = 1;
+        rp.m_Dependencies[2].m_Dst = dmGraphics::SUBPASS_EXTERNAL;
+
+        dmGraphics::VulkanCreateRenderPass(engine->m_GraphicsContext, m_Rendertarget, rp);
+    }
+
+    void Execute(EngineCtx* engine) override
+    {
+        dmGraphics::HTexture sub_pass_0_color = dmGraphics::GetRenderTargetTexture(m_Rendertarget, dmGraphics::BUFFER_TYPE_COLOR0_BIT);
+
+        dmGraphics::SetRenderTarget(engine->m_GraphicsContext, m_Rendertarget, 0);
+
+        dmGraphics::SetViewport(engine->m_GraphicsContext, 0, 0, 512, 512);
+
+        static uint8_t color_r = 0;
+        static uint8_t color_g = 80;
+        static uint8_t color_b = 140;
+        static uint8_t color_a = 255;
+
+        dmGraphics::Clear(engine->m_GraphicsContext, dmGraphics::BUFFER_TYPE_COLOR0_BIT,
+                                    (float)color_r,
+                                    (float)color_g,
+                                    (float)color_b,
+                                    (float)color_a,
+                                    1.0f, 0);
+
+        dmGraphics::VulkanNextRenderPass(engine->m_GraphicsContext, m_Rendertarget);
+
+        dmGraphics::EnableProgram(engine->m_GraphicsContext, m_ShaderProgram);
+        dmGraphics::DisableState(engine->m_GraphicsContext, dmGraphics::STATE_DEPTH_TEST);
+
+        dmGraphics::EnableTexture(engine->m_GraphicsContext, 0, 0, sub_pass_0_color);
+
+        dmGraphics::EnableVertexDeclaration(engine->m_GraphicsContext, m_VertexDeclaration, m_VertexBuffer);
+        dmGraphics::Draw(engine->m_GraphicsContext, dmGraphics::PRIMITIVE_TRIANGLES, 0, 6);
+
+        dmGraphics::SetRenderTarget(engine->m_GraphicsContext, 0, 0);
+    }
+};
+
+struct ComputeTest : ITest
```

Comment:
I added the test_app_vulkan.cpp, so that I could more easily debug stuff on Nintendo Switch.

I think generalising it is good, as that was the goal. (e.g. support OpenGL + PS5 too)
I just want us to not lose targets along the way.
E.g. will this new test app run on Switch when we merge it into the dswitch repo?

Reply:
I think it should, when Vulkan is being linked it will always gets picked first, so we shouldn't need any changes to this (but if we do, I'll make sure it still runs).

The changes I've made just enabled me to test the graphics API while building new features, and as such I'd like to be able to test any API for consitency. The alternative is to make separate API tests, but I think this is a better approach.